### PR TITLE
feat(barchart): reduce barchart creation verbosity

### PR DESCRIPTION
**Pull request by Emivvvvv**
implements a part of: https://github.com/ratatui/ratatui/issues/683

`new` and `with_label` constructors for `Bar`
`new` constructors for `BarGroup` and `BarChart`

### CODE REVIEW

File: examples/barchart-grouped.rs
Comment (orhun) [inline examples/barchart-grouped.rs:199]: `Bar::new("", /* */)` and `Bar::default()` are not the same here. One of them has `None` as the `label` and the other one is empty string. I would prefer the previous way of constructing it just to make sure we have the same behavior. (Same for the other parts where we use `""` in construction)

File: src/widgets/barchart/bar.rs
Comment (orhun) [inline src/widgets/barchart/bar.rs:238]: It is a bit unconventional to use tuples in `From` implementations. In other words, we don't have much examples of that in the codebase. Not a big problem though - I think it should be documented well.

File: src/widgets/barchart/bar_group.rs
Comment (orhun) [inline src/widgets/barchart/bar_group.rs:68]:
```suggestion
        self.bars = bars.into_iter().map(Into::into).collect();
```

File: src/widgets/barchart/bar_group.rs
Comment (orhun) [inline src/widgets/barchart/bar_group.rs:49]:
```suggestion
            bars: bars.into_iter().map(Into::into).collect(),
```

**Review by orhun — COMMENTED** [review]
Good refactor :)

I'm not sure about using empty strings in constructors. Maybe that `new` method should be renamed to `with_label` to make it clear that the label will be `Some(_)`

File: src/widgets/barchart/bar.rs
Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:238]: Where should I write the document about it? Top of the `From` implementation or somewhere else? I can try my best to document it, or removing it is also an option if others decide on this direction

**Comment by Emivvvvv** [comment]
> Good refactor :)
> 
> I'm not sure about using empty strings in constructors. Maybe that `new` method should be renamed to `with_label` to make it clear that the label will be `Some(_)`

Your point makes sense, but I'm also not sure. Do we have data that suggests how people use `Bar` generally? If they use it with labels, I think the label part should stay, but it should be removed if data suggests otherwise.

In both cases, I'm open to changing the implementation as you want :)

File: src/widgets/barchart/bar.rs
Comment (orhun) [reply src/widgets/barchart/bar.rs:238]: Let's wait for other maintainers' feedback first :)

**Review by joshka — COMMENTED** [review]
Thanks for the PR. I'm generally for the changes presented here, but have a questions to cover before going ahead.

**Comment by joshka** [comment]
> Your point makes sense, but I'm also not sure. Do we have data that suggests how people use `Bar` generally? If they use it with labels, I think the label part should stay, but it should be removed if data suggests otherwise.

I think this is worth questioning and finding a bit of an answer if it's easy.

Probably your best source for data on how things are used and what will break when doing breaking changes is going to be github searches. I generally add `lang:rust ratatui AND (NOT org:ratatui) AND (NOT is:fork)` to help filter out stuff which we wrote which we're not too concerned about breaking.

You may just have to look at a few random BarChart usages from random pages of results to get a feel for this.

File: src/widgets/barchart/bar.rs
Comment (orhun) [reply src/widgets/barchart/bar.rs:60]: Also, what do you think of renaming this to `with_label` which I mentioned above?

File: src/widgets/barchart/bar.rs
Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:94]: I also had this question while making the changes. I think splitting makes sense

File: src/widgets/barchart/bar.rs
Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:60]: I agree with both of your opinions. I think we should modify the `new` as it will only take the value and also have another impl `with_label` for more specific usages.

File: src/widgets/barchart/bar.rs
Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:251]: Can you give me a little helper blueprint on splitting this PR? I can reorganize the implementation and send separate PRs.

File: src/widgets/barchart.rs
Comment (joshka) [reply src/widgets/barchart.rs:146]: Rephrasing this - Do you think horizontal / vertical should be methods that set the direction, or methods that create a barchart object. The former sort of aligns with how e.g Line::centered() works, the latter aligns with how Layout::vertical() works. I lean towards the latter, but could see how the former would make sense too. Do you have any ideas about which makes the most sense?

File: src/widgets/barchart/bar.rs
Comment (joshka) [reply src/widgets/barchart/bar.rs:251]: How about these 3 PRs:
- `From<>` impl - breaking
- `Styled` impl
- new constructors and other things.

<details><summary>If your not super familiar with splitting up existing commits, at a high level the process is:</summary>


- create a new branch for the first PR 
- soft reset to the point where you branched from (there are 3 commits in this branch right now so that's `git reset HEAD~3`). This leaves the changes from the 3 commits in the index
- `git add --patch` the changed lines for the first commit
- commit that
- stash and test
- unstash
- cut a PR from that branch
- repeat

If in doubt, `git reflog` is your friend, and you should always be able to look at the hash of any commit you've written and update your current branch to that commit with `git reset --hard <somehash>`

</details>

Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:251]: > From<> impl - breaking
> Styled impl
> new constructors and other things.

These sound good to me, I will split this pr according to this guideline! And thanks a lot for the tips about splitting existing commits, I have never done this before so this is helpful.

Out of topic, I have an internship interview soon and  I'm currently studying for that. After I take the interview, on November 5, I'll start working on this PR.

Comment (joshka) [reply src/widgets/barchart/bar.rs:251]: No problem - take your time and good luck!

Comment (orhun) [reply src/widgets/barchart/bar.rs:251]: Best of luck man!

Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:251]: @joshka @orhun, thanks a lot for the good wishes :). It was an online assignment, and I decided to do it today, it went well. Let's see what happens next.

Back to work now, I assume you meant `Into<> impl` instead of `From<> impl` right?

File: src/widgets/barchart.rs
Comment (Emivvvvv) [reply src/widgets/barchart.rs:146]: In my opinion, setting `new()` to `Direction::Vertical` and also having both `vertical()` and `horizontal()` methods for creating a `BarChart` object makes the most sense.

File: src/widgets/barchart/bar.rs
Comment (Emivvvvv) [reply src/widgets/barchart/bar.rs:251]: @joshka I've done the changes however, I think I didn't do it with the way you wanted. 

To prevent conflicts, I opened 3 different branches that are built on top of each other (`Into<> impl - breaking` ->
`Styled impl` -> `new constructors and other things`). To explain, they all contains one commit, however `Styled impl` has the changes from the first branch, `Into<> impl - breaking`. Will this cause a problem? If it's not, I will open the PRs. If the PRs are merged in the correct order, there will be no problem.  But if you want the PR's completely separated from each other, I can separate them completely and then send PR's and resolve the conflicts later.

**Comment by Emivvvvv** [comment]
@orhun @joshka, I've split this PR into three as we discussed. I hope I did everything right.

I left examples unchanged. After all of the PRs have been accepted and merged, I'm willing to update unchanged examples or any extra work.

Here are the other PR's
feat(bar)!: update label and text_value to accept Into<>: https://github.com/ratatui/ratatui/pull/1471
feat(bar): impl Styled for Bar: https://github.com/ratatui/ratatui/pull/1476

**Review by joshka — APPROVED** [review]

**Comment by joshka** [comment]
Thanks for the PR @Emivvvvv

**Pull request merged**
